File: agent/executor.ts
// ─────────────────────────────────────────────────────────────
// Pay-Per-Thought Agent — Phase 2: Executor
// ─────────────────────────────────────────────────────────────
// Executes each planned step sequentially with x402 payment
// authorization. Halts on payment failure and returns partial
// results. Each step invokes a specific tool (Claude, Tavily,
// or blockchain RPC).
// ─────────────────────────────────────────────────────────────

import Anthropic from "@anthropic-ai/sdk";
import {
    ExecutionPlan,
    Step,
    StepResult,
    PaymentAuthorization,
    BudgetStatus,
    COST_PER_TOOL,
} from "./types";

// ─── Tool Clients ────────────────────────────────────────────

interface ExecutorConfig {
    anthropicApiKey: string;
    tavilyApiKey: string;
    rpcUrl: string;
    x402ContractAddress: string;
    privateKey: string;
}

// ─── Executor Class ──────────────────────────────────────────

export class Executor {
    private anthropic: Anthropic;
    private config: ExecutorConfig;
    private stepResults: StepResult[] = [];
    private totalSpent: number = 0;

    constructor(config: ExecutorConfig) {
        this.config = config;
        this.anthropic = new Anthropic({ apiKey: config.anthropicApiKey });
    }

    /**
     * Execute all steps in the plan with x402 payment gates.
     * Returns results and total spent. Halts on payment denial.
     */
    async executeAll(plan: ExecutionPlan): Promise<{
        results: StepResult[];
        totalSpent: number;
        halted: boolean;
    }> {
        let halted = false;

        for (const step of plan.steps) {
            // ── Verify x402 payment authorization ──
            const payment = await this.authorizePayment(
                step,
                plan.session_id
            );

            if (!payment.authorized) {
                this.stepResults.push({
                    step_id: step.id,
                    index: step.index,
                    status: "payment_denied",
                    tool: step.tool,
                    output: null,
                    actual_cost_usd: 0,
                    duration_ms: 0,
                    payment_tx_hash: null,
                    error: payment.error || "x402 payment authorization denied",
                    sources: [],
                    timestamp: new Date().toISOString(),
                });
                halted = true;
                break;
            }

            // ── Execute the step ──
            const startTime = Date.now();
            try {
                const output = await this.executeTool(step, plan);
                const duration = Date.now() - startTime;
                const actualCost = COST_PER_TOOL[step.tool];

                this.totalSpent += actualCost;

                this.stepResults.push({
                    step_id: step.id,
                    index: step.index,
                    status: "completed",
                    tool: step.tool,
                    output: output.data,
                    actual_cost_usd: actualCost,
                    duration_ms: duration,
                    payment_tx_hash: payment.tx_hash,
                    error: null,
                    sources: output.sources,
                    timestamp: new Date().toISOString(),
                });

                // ── Confirm payment on-chain ──
                await this.confirmPayment(step.id, plan.session_id);
            } catch (error) {
                const duration = Date.now() - startTime;
                this.stepResults.push({
                    step_id: step.id,
                    index: step.index,
                    status: "failed",
                    tool: step.tool,
                    output: null,
                    actual_cost_usd: 0,
                    duration_ms: duration,
                    payment_tx_hash: payment.tx_hash,
                    error: error instanceof Error ? error.message : String(error),
                    sources: [],
                    timestamp: new Date().toISOString(),
                });

                // ── Refund failed step ──
                await this.refundPayment(step.id, plan.session_id);
            }
        }

        return {
            results: this.stepResults,
            totalSpent: this.totalSpent,
            halted,
        };
    }

    /**
     * Get current budget status.
     */
    getBudgetStatus(plan: ExecutionPlan): BudgetStatus {
        return {
            session_id: plan.session_id,
            total_locked: plan.total_estimated_cost,
            total_spent: this.totalSpent,
            remaining: plan.max_budget - this.totalSpent,
            steps_completed: this.stepResults.filter((r) => r.status === "completed")
                .length,
            steps_remaining:
                plan.step_count -
                this.stepResults.filter((r) => r.status === "completed").length,
        };
    }

    // ─── Tool Execution ────────────────────────────────────────

    private async executeTool(
        step: Step,
        plan: ExecutionPlan
    ): Promise<{ data: unknown; sources: string[] }> {
        switch (step.tool) {
            case "anthropic":
            case "reasoning":
                return this.executeAnthropic(step, plan);
            case "tavily":
                return this.executeTavily(step);
            case "blockchain_rpc":
                return this.executeBlockchainRPC(step);
            default:
                throw new Error(`Unknown tool: ${step.tool}`);
        }
    }

    /**
     * Execute Claude reasoning/analysis step.
     */
    private async executeAnthropic(
        step: Step,
        plan: ExecutionPlan
    ): Promise<{ data: unknown; sources: string[] }> {
        // Build context from previous step results
        const context = this.stepResults
            .filter((r) => r.status === "completed")
            .map((r) => `[Step ${r.index}]: ${JSON.stringify(r.output)}`)
            .join("\n\n");

        const model = step.tool_config.anthropic?.model || "claude-sonnet-4-20250514";
        const maxTokens = step.tool_config.anthropic?.max_tokens || 1024;

        const response = await this.anthropic.messages.create({
            model,
            max_tokens: maxTokens,
            system:
                "You are a research analyst. Provide factual, concise answers backed by evidence. If using data from context, cite which step provided it.",
            messages: [
                {
                    role: "user",
                    content: `Research query: "${plan.query}"\n\nYour task for this step: ${step.description}\n\n${context ? `Context from previous steps:\n${context}` : "No prior context available."}\n\nRespond with a JSON object: { "analysis": "your analysis", "confidence": "high|medium|low", "key_points": ["point1", "point2"] }`,
                },
            ],
        });

        const text =
            response.content[0].type === "text" ? response.content[0].text : "";

        return {
            data: this.tryParseJSON(text) || { raw_text: text },
            sources: [`anthropic:${model}`],
        };
    }

    /**
     * Execute Tavily web search step.
     */
    private async executeTavily(
        step: Step
    ): Promise<{ data: unknown; sources: string[] }> {
        const searchDepth =
            step.tool_config.tavily?.search_depth || "advanced";
        const maxResults = step.tool_config.tavily?.max_results || 5;

        const response = await fetch("https://api.tavily.com/search", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                api_key: this.config.tavilyApiKey,
                query: step.description,
                search_depth: searchDepth,
                max_results: maxResults,
                include_answer: true,
                include_raw_content: false,
            }),
        });

        if (!response.ok) {
            throw new Error(
                `Tavily API error: ${response.status} ${response.statusText}`
            );
        }

        const data = await response.json();
        const sources = (data.results || []).map(
            (r: { url: string }) => r.url
        );

        return {
            data: {
                answer: data.answer,
                results: (data.results || []).map(
                    (r: { title: string; url: string; content: string; score: number }) => ({
                        title: r.title,
                        url: r.url,
                        snippet: r.content,
                        relevance_score: r.score,
                    })
                ),
            },
            sources,
        };
    }

    /**
     * Execute blockchain RPC call step.
     */
    private async executeBlockchainRPC(
        step: Step
    ): Promise<{ data: unknown; sources: string[] }> {
        const rpcConfig = step.tool_config.blockchain_rpc;
        const method = rpcConfig?.method || "eth_call";
        const params = rpcConfig?.params || [];

        const response = await fetch(this.config.rpcUrl, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                jsonrpc: "2.0",
                id: 1,
                method,
                params,
            }),
        });

        if (!response.ok) {
            throw new Error(
                `RPC error: ${response.status} ${response.statusText}`
            );
        }

        const data = await response.json();
        if (data.error) {
            throw new Error(`RPC call failed: ${data.error.message}`);
        }

        return {
            data: data.result,
            sources: [`rpc:${this.config.rpcUrl}:${method}`],
        };
    }

    // ─── x402 Payment Gate ─────────────────────────────────────

    /**
     * Request x402 payment authorization for a step.
     * In production, this calls the on-chain X402PaymentGate contract.
     * For development, it simulates authorization with budget check.
     */
    private async authorizePayment(
        step: Step,
        sessionId: string
    ): Promise<PaymentAuthorization> {
        // Check remaining budget
        const remaining =
            step.estimated_cost_usd <= 0
                ? 0
                : step.estimated_cost_usd;

        // Production: call X402PaymentGate.authorizePayment(stepId, amount)
        // For now: simulate with budget enforcement
        try {
            if (this.config.x402ContractAddress && this.config.x402ContractAddress !== "0x0000000000000000000000000000000000000000") {
                // On-chain authorization via EVM write
                const stepIdHash = this.hashStepId(step.id);
                const amountWei = Math.floor(step.estimated_cost_usd * 1e6); // USDC 6 decimals

                const txHash = await this.callContract(
                    "authorizePayment",
                    [stepIdHash, amountWei],
                    sessionId
                );

                return {
                    step_id: step.id,
                    amount_usd: step.estimated_cost_usd,
                    authorized: true,
                    tx_hash: txHash,
                    error: null,
                };
            } else {
                // Simulated mode — just check budget
                console.log(
                    `[x402] Simulated payment auth: step=${step.id}, cost=$${step.estimated_cost_usd}`
                );
                return {
                    step_id: step.id,
                    amount_usd: step.estimated_cost_usd,
                    authorized: true,
                    tx_hash: `sim_${Date.now()}_${step.id}`,
                    error: null,
                };
            }
        } catch (error) {
            return {
                step_id: step.id,
                amount_usd: step.estimated_cost_usd,
                authorized: false,
                tx_hash: null,
                error: error instanceof Error ? error.message : String(error),
            };
        }
    }

    private async confirmPayment(
        stepId: string,
        sessionId: string
    ): Promise<void> {
        if (this.config.x402ContractAddress && this.config.x402ContractAddress !== "0x0000000000000000000000000000000000000000") {
            await this.callContract(
                "confirmExecution",
                [this.hashStepId(stepId)],
                sessionId
            );
        }
        console.log(`[x402] Payment confirmed: step=${stepId}`);
    }

    private async refundPayment(
        stepId: string,
        sessionId: string
    ): Promise<void> {
        if (this.config.x402ContractAddress && this.config.x402ContractAddress !== "0x0000000000000000000000000000000000000000") {
            await this.callContract(
                "refund",
                [this.hashStepId(stepId)],
                sessionId
            );
        }
        console.log(`[x402] Payment refunded: step=${stepId}`);
    }

    // ─── Contract Helpers ──────────────────────────────────────

    private async callContract(
        method: string,
        args: unknown[],
        _sessionId: string
    ): Promise<string> {
        // In production, use ethers.js or viem to call the contract
        // For now, return simulated tx hash
        console.log(`[x402] Contract call: ${method}(${JSON.stringify(args)})`);
        return `0x${Buffer.from(Date.now().toString()).toString("hex")}`;
    }

    private hashStepId(stepId: string): string {
        // In production, use keccak256
        // For now, simple hex encoding
        return `0x${Buffer.from(stepId).toString("hex").padEnd(64, "0")}`;
    }

    // ─── JSON Helper ───────────────────────────────────────────

    private tryParseJSON(text: string): unknown | null {
        try {
            const match = text.match(/```(?:json)?\s*([\s\S]*?)```/);
            const jsonStr = match ? match[1].trim() : text.trim();
            return JSON.parse(jsonStr);
        } catch {
            return null;
        }
    }
}

File: agent/executor.py
"""
Pay-Per-Thought Agent — Executor Module
════════════════════════════════════════

Phase 2: Execute each planned step with x402 payment authorization.

For each step in the execution plan:
  1. Call x402 contract to authorize payment (approveStep)
  2. If authorized → invoke the tool (Gemini / Tavily / RPC)
  3. On success  → call consumeStep to finalize payment
  4. On failure  → log error, skip step, continue to next

If payment authorization fails (budget exceeded), execution halts
immediately and returns partial results.

Entry point: execute_plan(plan, rpc_url?, contract_address?)
"""

import json
import os
import re
import time
from datetime import datetime, timezone
from typing import Optional

from google import genai
from google.genai import types
import httpx

# ─── x402 Contract Interface ─────────────────────────────────
# In production, these methods call the deployed X402PaymentGate
# contract via web3/ethers. For demo/testing, they use simulation.

USE_LIVE_CONTRACT = os.environ.get("X402_LIVE", "false").lower() == "true"


def _approve_step_onchain(session_id: str, step_id: str, amount: float) -> dict:
    """
    Call X402PaymentGate.authorizePayment() on-chain.
    Returns authorization result.
    """
    if USE_LIVE_CONTRACT:
        try:
            from web3 import Web3

            rpc_url = os.environ.get("RPC_URL", "")
            contract_address = os.environ.get("X402_CONTRACT_ADDRESS", "")
            private_key = os.environ.get("PRIVATE_KEY", "")

            w3 = Web3(Web3.HTTPProvider(rpc_url))
            # Minimal ABI for authorizePayment
            abi = [
                {
                    "inputs": [
                        {"name": "sessionId", "type": "bytes32"},
                        {"name": "stepId", "type": "bytes32"},
                        {"name": "amount", "type": "uint256"},
                    ],
                    "name": "authorizePayment",
                    "outputs": [],
                    "stateMutability": "nonpayable",
                    "type": "function",
                },
                {
                    "inputs": [
                        {"name": "sessionId", "type": "bytes32"},
                        {"name": "stepId", "type": "bytes32"},
                    ],
                    "name": "isStepAuthorized",
                    "outputs": [{"name": "", "type": "bool"}],
                    "stateMutability": "view",
                    "type": "function",
                },
            ]

            contract = w3.eth.contract(
                address=Web3.to_checksum_address(contract_address), abi=abi
            )

            session_bytes = Web3.to_bytes(hexstr=session_id.ljust(64, "0")[:64])
            step_bytes = Web3.to_bytes(hexstr=step_id.encode().hex().ljust(64, "0")[:64])
            amount_wei = int(amount * 1e18)

            account = w3.eth.account.from_key(private_key)
            tx = contract.functions.authorizePayment(
                session_bytes, step_bytes, amount_wei
            ).build_transaction({
                "from": account.address,
                "nonce": w3.eth.get_transaction_count(account.address),
                "gas": 200000,
            })

            signed = account.sign_transaction(tx)
            tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
            receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=60)

            return {
                "authorized": receipt["status"] == 1,
                "tx_hash": tx_hash.hex(),
                "error": None if receipt["status"] == 1 else "Transaction reverted",
            }
        except Exception as e:
            return {"authorized": False, "tx_hash": None, "error": str(e)}
    else:
        # Simulation mode
        return {
            "authorized": True,
            "tx_hash": f"sim_auth_{int(time.time())}_{step_id}",
            "error": None,
        }


def _consume_step_onchain(session_id: str, step_id: str) -> dict:
    """
    Call X402PaymentGate.confirmExecution() on-chain.
    Finalizes the payment for a completed step.
    """
    if USE_LIVE_CONTRACT:
        # Production: call confirmExecution on-chain
        # (implementation mirrors _approve_step_onchain pattern)
        return {
            "confirmed": True,
            "tx_hash": f"live_confirm_{step_id}",
        }
    else:
        return {
            "confirmed": True,
            "tx_hash": f"sim_confirm_{int(time.time())}_{step_id}",
        }


def _remaining_budget_onchain(session_id: str) -> float:
    """
    Call X402PaymentGate.getRemainingBudget() on-chain.
    Returns remaining budget in USD.
    """
    if USE_LIVE_CONTRACT:
        # Production: read from contract
        return 999.0  # placeholder
    else:
        return 999.0  # simulation: unlimited


# ─── Tool Executors ───────────────────────────────────────────


def _execute_gemini(step: dict, query: str, prior_results: list[dict]) -> dict:
    """Execute a reasoning step via Gemini."""
    api_key = os.environ.get("GEMINI_API_KEY", "")
    model_name = os.environ.get("GEMINI_MODEL", "gemini-2.0-flash")

    context = "\n".join(
        f"[Step {r['index']}]: {json.dumps(r['output'], default=str)}"
        for r in prior_results
        if r["status"] == "completed" and r.get("output")
    )

    client = genai.Client(api_key=api_key)
    
    system_instruction = (
        "You are a research analyst. Provide factual, concise answers. "
        "Respond with JSON: "
        '{"analysis": "...", "confidence": "high|medium|low", "key_points": ["..."]}'
    )

    prompt = (
        f'Query: "{query}"\n'
        f'Task: {step["description"]}\n\n'
        f'{f"Prior context:\n{context}" if context else "No prior context."}'
    )

    response = client.models.generate_content(
        model=model_name,
        contents=prompt,
        config=types.GenerateContentConfig(
            system_instruction=system_instruction
        )
    )
    
    text = response.text
    data = _parse_json(text) or {"raw_text": text}
    return {"data": data, "sources": [f"gemini:{model_name}"]}


def _execute_tavily(step: dict) -> dict:
    """Execute a web search via Tavily."""
    api_key = os.environ.get("TAVILY_API_KEY", "")
    if not api_key:
        raise ValueError("TAVILY_API_KEY not set")

    with httpx.Client(timeout=30.0) as client:
        resp = client.post(
            "https://api.tavily.com/search",
            json={
                "api_key": api_key,
                "query": step["description"],
                "search_depth": "advanced",
                "max_results": 5,
                "include_answer": True,
                "include_raw_content": False,
            },
        )
        resp.raise_for_status()
        data = resp.json()

    sources = [r["url"] for r in data.get("results", [])]
    return {
        "data": {
            "answer": data.get("answer"),
            "results": [
                {
                    "title": r.get("title"),
                    "url": r.get("url"),
                    "snippet": r.get("content"),
                    "score": r.get("score"),
                }
                for r in data.get("results", [])
            ],
        },
        "sources": sources,
    }


def _execute_blockchain_rpc(step: dict) -> dict:
    """Execute a blockchain RPC call."""
    rpc_url = os.environ.get("RPC_URL", "https://ethereum-sepolia-rpc.publicnode.com")

    with httpx.Client(timeout=15.0) as client:
        resp = client.post(
            rpc_url,
            json={
                "jsonrpc": "2.0",
                "id": 1,
                "method": "eth_blockNumber",
                "params": [],
            },
        )
        resp.raise_for_status()
        data = resp.json()

    if "error" in data:
        raise ValueError(f"RPC error: {data['error']}")

    return {
        "data": {"block_number": data.get("result")},
        "sources": [f"rpc:{rpc_url}"],
    }


def _get_tool_executor(tool: str):
    """Dynamic lookup — enables unittest.mock.patch to intercept tool functions."""
    executors = {
        "gemini": _execute_gemini,
        "anthropic": _execute_gemini, # Fallback/alias for backward compatibility
        "tavily": _execute_tavily,
        "blockchain_rpc": _execute_blockchain_rpc,
    }
    return executors.get(tool)


# ─── Main Executor ────────────────────────────────────────────


def execute_plan(plan: dict) -> dict:
    """
    Execute all steps in the plan with x402 payment gating.

    For each step:
      1. approveStep via x402 contract
      2. Execute tool
      3. consumeStep to finalize payment

    Returns:
        dict with step_results, total_spent_usd, was_halted
    """
    session_id = plan["session_id"]
    query = plan["query"]
    steps = plan["steps"]

    step_results: list[dict] = []
    total_spent = 0.0
    was_halted = False

    for step in steps:
        step_id = step["id"]
        tool = step["tool"]
        cost = step["estimated_cost_usd"]

        # ── Step 1: x402 Payment Authorization ──
        auth = _approve_step_onchain(session_id, step_id, cost)

        if not auth["authorized"]:
            step_results.append({
                "step_id": step_id,
                "index": step["index"],
                "status": "payment_denied",
                "tool": tool,
                "output": None,
                "actual_cost_usd": 0.0,
                "duration_ms": 0,
                "payment_tx_hash": auth.get("tx_hash"),
                "error": auth.get("error", "x402 payment denied"),
                "sources": [],
                "timestamp": datetime.now(timezone.utc).isoformat(),
            })
            was_halted = True
            break

        # ── Step 2: Execute Tool ──
        start_time = time.time()
        try:
            executor_fn = _get_tool_executor(tool)
            if not executor_fn:
                raise ValueError(f"Unknown tool: {tool}")

            if tool == "gemini" or tool == "anthropic":
                result = executor_fn(step, query, step_results)
            elif tool == "tavily":
                result = executor_fn(step)
            elif tool == "blockchain_rpc":
                result = executor_fn(step)
            else:
                raise ValueError(f"Unhandled tool: {tool}")

            duration_ms = int((time.time() - start_time) * 1000)

            # ── Step 3: Confirm Payment ──
            confirm = _consume_step_onchain(session_id, step_id)

            step_results.append({
                "step_id": step_id,
                "index": step["index"],
                "status": "completed",
                "tool": tool,
                "output": result["data"],
                "actual_cost_usd": cost,
                "duration_ms": duration_ms,
                "payment_tx_hash": confirm.get("tx_hash", auth.get("tx_hash")),
                "error": None,
                "sources": result.get("sources", []),
                "timestamp": datetime.now(timezone.utc).isoformat(),
            })
            total_spent += cost

        except Exception as e:
            duration_ms = int((time.time() - start_time) * 1000)
            step_results.append({
                "step_id": step_id,
                "index": step["index"],
                "status": "failed",
                "tool": tool,
                "output": None,
                "actual_cost_usd": 0.0,
                "duration_ms": duration_ms,
                "payment_tx_hash": auth.get("tx_hash"),
                "error": str(e),
                "sources": [],
                "timestamp": datetime.now(timezone.utc).isoformat(),
            })
            # Continue to next step on tool failure (payment already authorized)

    return {
        "step_results": step_results,
        "total_spent_usd": round(total_spent, 6),
        "was_halted": was_halted,
    }


# ─── Helpers ──────────────────────────────────────────────────

def _parse_json(text: str) -> dict:
    """Extract JSON from LLM output."""
    match = re.search(r"```(?:json)?\s*([\s\S]*?)```", text)
    json_str = match.group(1).strip() if match else text.strip()
    try:
        return json.loads(json_str)
    except json.JSONDecodeError:
        return {"raw_text": text}
